python/lsst/pipe/tasks/dcrAssembleCoadd.py

```diff
@@ -0,0 +1,781 @@
+                shift = (-dcr.dy, -dcr.dx)
+            else:
+                shift = (dcr.dy, dcr.dx)
+            # Shift each of image, mask, and variance
+            if bbox is None:
+                result = maskedImage.clone()
+            else:
+                result = maskedImage[bbox, afwImage.PARENT].clone()
+            mask = result.getMask()
+            mask &= ~mask.getPlaneBitMask("CLIPPED")
+            srcImage = result.getImage().getArray()
+            scrVariance = result.getVariance().getArray()
+            nanInds = numpy.isnan(srcImage)
+            mask.getArray()[nanInds] = mask.getPlaneBitMask("NO_DATA")
+            result.setMask(DcrAssembleCoaddTask.shiftMask(mask, dcr, useInverse=useInverse))
+
+            srcImage[nanInds] = 0.
+            # scrVariance[nanInds] = numpy.inf
+            retImage = scipy.ndimage.interpolation.shift(srcImage, shift)
+            result.getImage().getArray()[:, :] = retImage
+            retVariance = scipy.ndimage.interpolation.shift(scrVariance, shift)
+            result.getVariance().getArray()[:, :] = retVariance
+        return result
+
+    @staticmethod
+    def conditionDcrModel(oldDcrModels, newDcrModels, bbox, gain=1.):
+        """! Average the current solution with the solution from the last iteration to reduce oscillations.
+
+        @param[in] oldDcrModels: A list of masked images, each containing
+                                the model for one subfilter from the previous iteration.
+        @param[in, out] newDcrModels: A list of masked images, each containing
+                                      the model for one subfilter from the current iteration.
+        @param[in] bbox: Sub-region of the coadd
+        @param[in] gain: Additional weight to apply to the model from the current iteration.
+        """
+        for oldModel, newModel in zip(oldDcrModels, newDcrModels):
+            # The DcrModels are MaskedImages, which only support in-place operations.
+            newModel *= gain
+            newModel += oldModel[bbox, afwImage.PARENT]
+            newModel.getImage().getArray()[:, :] /= 1. + gain
+            newModel.getVariance().getArray()[:, :] /= 1. + gain
+
+    @staticmethod
+    def dcrShiftCalculate(visitInfo, wcs, lambdaEff, filterWidth, dcrNSubbands):
+        """! Calculate the shift in pixels of an exposure due to DCR.
+
+        @param[in] visitInfo: lsst.afw.image.VisitInfo for the exposure.
+        @param[in] wcs: lsst.afw.geom.skyWcs.skyWcs.SkyWcs wcs for the exposure.
+        @param[in] lambdaEff: Effective center wavelength of the filter, in nm.
+        @param[in] filterWidth: Width of the filter, in nm.
+        @param[in] dcrNSubbands: Number of subfilters to divide the full band into.
+
+        @return named tuple `dcr` containing the shift due to DCR, in pixels.
+        """
+        rotation = DcrAssembleCoaddTask.calculateRotationAngle(visitInfo, wcs)
+
+        dcr = namedtuple("dcr", ["dx", "dy"])
```

Comment:
We have a class that contains dx,dy tuples: `Extent2D`. Please use that instead.